src/uct/tcp/tcp_ep.c

```diff
@@ -619,24 +625,22 @@ static inline unsigned uct_tcp_ep_recv(uct_tcp_ep_t *ep, size_t recv_length)
 static unsigned uct_tcp_ep_progress_data_tx(uct_tcp_ep_t *ep)
 {
     unsigned count = 0;
-    size_t sent_length;
-    int ret;
+    ssize_t ret;
 
     ucs_trace_func("ep=%p", ep);
 
     if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
         ret = (!(ep->ctx_caps & UCS_BIT(UCT_TCP_EP_CTX_TYPE_ZCOPY_TX)) ?
-               uct_tcp_ep_send(ep, &sent_length) :
-               uct_tcp_ep_sendv(ep, &sent_length));
+               uct_tcp_ep_send(ep) : uct_tcp_ep_sendv(ep));
         if (ucs_unlikely(UCS_STATUS_IS_ERR(ret))) {
             uct_tcp_ep_handle_disconnected(ep, &ep->tx);
             return 0;
```

Comment:
return 1

Reply:
`0` means that call progressed no messaged

Reply:
failure needs to be propagated to upper level well as successful messaging

Reply:
`uct_tcp_ep_progress_data_tx` is called from `uct_iface_progress()` that doesn't support returning an error to a user

Reply:
> anyway we have to return `> 0` value from `ucp_worker_progress` if something is handled, including failures.

IB returns 0 in case of failure:

https://github.com/openucx/ucx/blob/079359fcc9011efe547833c36072cf203a1d5d90/src/uct/ib/rc/verbs/rc_verbs_impl.h#L118

Reply:
IB returns >0 on TX
https://github.com/openucx/ucx/blob/079359fcc9011efe547833c36072cf203a1d5d90/src/uct/ib/rc/verbs/rc_verbs_iface.c#L119

failure on RX side is fatal
https://github.com/openucx/ucx/blob/079359fcc9011efe547833c36072cf203a1d5d90/src/uct/ib/base/ib_device.h#L303

Reply:
> IB returns >0 on TX

no, it returns `0` from foreach in case of failure for TX:
https://github.com/openucx/ucx/blob/079359fcc9011efe547833c36072cf203a1d5d90/src/uct/ib/rc/verbs/rc_verbs.h#L17

Reply:
No, this case is also fatal

Reply:
yes
the same for TCP. if `send()`/`sendmsg()` returns `-1` and `errno != EAGAIN/EWOULDBLOCK/EINTR`, this is fatal error

Reply:
so you call `uct_tcp_ep_handle_disconnected` on `EAGAIN/EWOULDBLOCK/EINTR`?

Reply:
no, because this is `UCS_ERR_NO_PROGRESS`, i.e. `uct_tcp_ep_send` will return it

Reply:
Ok
1) when this branch is executed?
2) you have to return 1 here because of handling disconnect event

Reply:
I got your idea, will be fixed

Reply:
> 1. when this branch is executed?

when an error occurred, i.e. neither `UCS_OK` nor `UCS_ERR_NO_PROGRESS` returned form `ucs_socket_sendv_nb()`